pkg/kubectl/cmd/util/openapi/validation/types.go

```diff
@@ -0,0 +1,265 @@
+		sortedKeys = append(sortedKeys, key)
+	}
+	sort.Strings(sortedKeys)
+	return sortedKeys
+}
+
+var _ ValidationItem = &mapItem{}
+
+func (mi *mapItem) VisitPrimitive(p *openapi.Primitive) {
+	mi.AddValidationError(InvalidTypeError{Path: p.GetPath().String(), Expected: p.Type, Actual: "map"})
+}
+
+func (mi *mapItem) VisitArray(a *openapi.Array) {
+	mi.AddValidationError(InvalidTypeError{Path: a.GetPath().String(), Expected: "array", Actual: "map"})
+}
+
+func (mi *mapItem) VisitMap(m *openapi.Map) {
+
+	for _, key := range mi.sortedKeys() {
+		item, err := itemFactory(mi.Path().FieldPath(key), mi.Map[key])
+		if err != nil {
+			mi.AddError(err)
+			continue
+		}
+		m.SubType.Accept(item)
+		mi.CopyErrors(item.Errors())
+	}
+}
+
+func (mi *mapItem) VisitKind(k *openapi.Kind) {
+	// Verify each sub-field.
+	for _, key := range mi.sortedKeys() {
+		item, err := itemFactory(mi.Path().FieldPath(key), mi.Map[key])
+		if err != nil {
+			mi.AddError(err)
+			continue
+		}
+		if _, ok := k.Fields[key]; !ok {
+			mi.AddValidationError(UnknownFieldError{Path: k.GetPath().String(), Field: key})
+			continue
+		}
+		k.Fields[key].Accept(item)
+		mi.CopyErrors(item.Errors())
+	}
+
+	// Verify that all required fields are present.
+	for _, required := range k.RequiredFields {
+		if _, ok := mi.Map[required]; !ok {
+			mi.AddValidationError(MissingRequiredFieldError{Path: k.GetPath().String(), Field: required})
+		}
+	}
+}
+
+// arrayItem represents a yaml array.
+type arrayItem struct {
+	baseItem
+
+	Array []interface{}
+}
+
+var _ ValidationItem = &arrayItem{}
+
+func (ai *arrayItem) VisitPrimitive(p *openapi.Primitive) {
+	ai.AddValidationError(InvalidTypeError{Path: p.GetPath().String(), Expected: p.Type, Actual: "array"})
+}
+
+func (ai *arrayItem) VisitArray(a *openapi.Array) {
+	for i, v := range ai.Array {
+		item, err := itemFactory(ai.Path().ArrayPath(i), v)
+		if err != nil {
+			ai.AddError(err)
+			continue
+		}
+		a.SubType.Accept(item)
+		ai.CopyErrors(item.Errors())
+	}
+}
+
+func (ai *arrayItem) VisitMap(m *openapi.Map) {
+	ai.AddValidationError(InvalidTypeError{Path: m.GetPath().String(), Expected: "array", Actual: "map"})
+}
+
+func (ai *arrayItem) VisitKind(k *openapi.Kind) {
+	ai.AddValidationError(InvalidTypeError{Path: k.GetPath().String(), Expected: "array", Actual: "map"})
+}
+
+// primitiveItem represents a yaml value.
+type primitiveItem struct {
+	baseItem
+
+	Value interface{}
+	Kind  string
+}
+
+var _ ValidationItem = &primitiveItem{}
+
+func (pi *primitiveItem) VisitPrimitive(p *openapi.Primitive) {
+	// Some types of primitives can match more than one (a number
+	// can be a string, but not the other way around).
+	// Return from the switch if we have a valid possible type conversion
+	switch p.Type {
+	case openapi.Boolean:
+		switch pi.Kind {
+		case openapi.Boolean:
+			return
+		}
+	case openapi.Integer:
+		switch pi.Kind {
+		case openapi.Integer, openapi.Number:
+			return
+		}
+	case openapi.Number:
```

Comment:
This is mapping to primitive types as defined in the [openapi spec](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md#data-types)